scrapy_poet/page_input_providers.py

```diff
@@ -263,3 +282,77 @@ class ResponseUrlProvider(PageObjectInputProvider):
     def __call__(self, to_provide: Set[Callable], response: Response):
         """Builds a ``ResponseUrl`` instance using a Scrapy ``Response``."""
         return [ResponseUrl(url=response.url)]
+
+
+class ItemProvider(PageObjectInputProvider):
+
+    name = "item"
+
+    def __init__(self, injector):
+        super().__init__(injector)
+        self.registry = self.injector.overrides_registry
+
+        # The key that's used here is the ``scrapy.Request`` instance to ensure
+        # that the cached instances under it are properly garbage collected
+        # after processing such request.
+        self._cached_instances = WeakKeyDictionary()
+
+    def provided_classes(self, cls):
+        """If the item is in any of the ``to_return`` in the rules, then it can
+        definitely provide by using the corresponding page object in ``use``.
+        """
+        return isclass(cls) and self.registry.search(to_return=cls)
```

Comment:
Won't it be expensive to call self.registry.search all the time? It's now O(N), going through all the defined rules. And it also seems that provided_classes is called for each request, if I'm not mistaken.

Reply:
Good call!

I've updated the PR https://github.com/scrapinghub/scrapy-poet/pull/103 instead since it's directly using the new `web_poet.RulesRegistry` which offers `O(1)` look-up time. https://github.com/scrapinghub/scrapy-poet/pull/103/commits/3c6fdaec67cf1f386de46ddc1af34671c1ef5546

Reply:
It'd be better not to use a private method. We can either introduce another method, or (probably better) optimize .search for this use case.

Reply:
> or (probably better) optimize .search for this use case.

That's indeed much better 👍

Optimized the `RulesRegistry.search()` method in https://github.com/scrapinghub/web-poet/pull/112/commits/05a832b82beae4242a3facc1920992cee91e00d2 to have a faster look-up when `to_return` or `instead_of` are used.